File: Go/ExampleTests/common/address_string.go
package common

type AddressString struct {
	Street string
	City string
	State string
	ZIP string
}

func NewAddressStringFromSlice(v []string) AddressString {
	s := AddressString{}
	if len(v) > 0 { s.Street = v[0] }
	if len(v) > 1 { s.City = v[1] }
	if len(v) > 2 { s.State = v[2] }
	if len(v) > 3 { s.ZIP = v[3] }
	return s
}

// NewAddressStringFromText builds from the text form, e.g. Money as "25 USD".
func NewAddressStringFromText(text string) AddressString {
	parts := RequireTokens(text, 4, "Address")
	return AddressString{
		Street: parts[0],
		City: parts[1],
		State: parts[2],
		ZIP: parts[3],
	}
}

func (s AddressString) String() string {
	return Token(s.Street) + " " + Token(s.City) + " " + Token(s.State) + " " + Token(s.ZIP)
}

func (s AddressString) Equals(o AddressString) bool {
	return DNCEqual(s.Street, o.Street) &&
		DNCEqual(s.City, o.City) &&
		DNCEqual(s.State, o.State) &&
		DNCEqual(s.ZIP, o.ZIP)
}

func EqualAddressStringSlices(a, b []AddressString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/address_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct AddressString {
    pub street: String,
    pub city: String,
    pub state: String,
    pub zip: String,
}

impl AddressString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            street: v.get(0).copied().unwrap_or("").to_string(),
            city: v.get(1).copied().unwrap_or("").to_string(),
            state: v.get(2).copied().unwrap_or("").to_string(),
            zip: v.get(3).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 4, "Address");
        Self {
            street: parts[0].clone(),
            city: parts[1].clone(),
            state: parts[2].clone(),
            zip: parts[3].clone(),
        }
    }
}

impl std::fmt::Display for AddressString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.street), crate::common::tokens::token(&self.city), crate::common::tokens::token(&self.state), crate::common::tokens::token(&self.zip)].join(" "))
    }
}

impl PartialEq for AddressString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.street, &other.street)
            && crate::common::dnc_equal(&self.city, &other.city)
            && crate::common::dnc_equal(&self.state, &other.state)
            && crate::common::dnc_equal(&self.zip, &other.zip)
    }
}
